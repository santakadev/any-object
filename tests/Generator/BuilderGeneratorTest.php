<?php

namespace Santakadev\AnyObject\Tests\Generator;

use ApprovalTests\Approvals;
use Santakadev\AnyObject\Tests\Generator\Generated\MoneyBuilder;
use Santakadev\AnyObject\Tests\Generator\Generated\ProductBuilder;
use Santakadev\AnyObject\Tests\Generator\Generated\ProductPriceBuilder;
use Santakadev\AnyObject\Tests\Generator\Generated\QuantityBuilder;
use Santakadev\AnyObject\Tests\TestData\ComplexType\Money\Amount;
use Santakadev\AnyObject\Tests\TestData\ComplexType\Product;
use Santakadev\AnyObject\Tests\TestData\ComplexType\ProductId;
use Santakadev\AnyObject\Tests\TestData\ComplexType\ProductName;
use Santakadev\AnyObject\Tests\TestData\ComplexType\ProductPrice;
use Santakadev\AnyObject\Tests\TestData\ComplexType\Quantity;

class BuilderGeneratorTest extends BuilderGeneratorTestCase
{
    public function test_builder(): void
    {
        $this->generateBuilderFor(Quantity::class);
        $text = $this->readGeneratedAnyFileFor(Quantity::class);
        Approvals::verifyString($text);

        $quantity = QuantityBuilder::create()
            ->withValue(3)
            ->build();

        $this->assertEquals(3, $quantity->value);
    }

    public function test_complex_builder(): void
    {
        $this->generateBuilderFor(Product::class);
        $text = $this->readGeneratedAnyFileFor(Product::class);
        Approvals::verifyString($text);

        $money = MoneyBuilder::create()
            ->withAmount(new Amount(5))
            ->build();
        $productPrice = new ProductPrice($money);

        $product = ProductBuilder::create()
            ->withId(new ProductId('id'))
            ->withName(new ProductName('name'))
            ->withPrice($productPrice)
            ->build();

        $this->assertEquals('id', $product->id()->value);
        $this->assertEquals('name', $product->name()->value);
        $this->assertEquals(5, $product->price()->value->amount->value);
    }
}
