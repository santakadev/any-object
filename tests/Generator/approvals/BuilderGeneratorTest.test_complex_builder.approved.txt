<?php

namespace Santakadev\AnyObject\Tests\Generator\Generated;

use Faker\Factory;
use Santakadev\AnyObject\Tests\TestData\ComplexType\Product;
use Santakadev\AnyObject\Tests\TestData\ComplexType\ProductId;
use Santakadev\AnyObject\Tests\TestData\ComplexType\ProductName;
use Santakadev\AnyObject\Tests\TestData\ComplexType\ProductPrice;

final class ProductBuilder
{
    private function __construct(private ProductId $id, private ProductName $name, private ProductPrice $price)
    {
    }
    public static function create() : self
    {
        $faker = Factory::create();
        $id = ProductIdBuilder::create()->build();
        $name = ProductNameBuilder::create()->build();
        $price = ProductPriceBuilder::create()->build();
        return new self($id, $name, $price);
    }
    public function withId(ProductId $id) : self
    {
        $this->id = $id;
        return $this;
    }
    public function withName(ProductName $name) : self
    {
        $this->name = $name;
        return $this;
    }
    public function withPrice(ProductPrice $price) : self
    {
        $this->price = $price;
        return $this;
    }
    public function build() : Product
    {
        return new Product($this->id, $this->name, $this->price);
    }
}
